### PR TITLE
Enable boosted tau reco to run at miniAOD

### PR DESCRIPTION
**Pull request by mbluj**
#### PR description:

This development enables boosted tau pair reco to be run on top of miniAOD. To achieve this goal the following changes were introduced:
* BoostedTauSeedProducer was templated to enable miniAOD inputs,
* Tool to adapt tau sequences was modified (functions grouped into class, possibility to use postfix added, boosted tau pair reco enabled, some cleaning).

These changes do not modifies standard workflows.

Note: boosted tau pair reco with miniAOD inputs compared to its counterpart with AOD inputs shows similar differences as observed for standard tau reco.

#### PR validation:

A MiniAOD production workflow and a custom tau-reco-at-mini one successfully tested.

### CODE REVIEW

**Comment by mbluj** [comment]
Development branch was rebased to master and conflicts solved.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by mbluj** [comment]
There is a number of failing workflows, but to me those are not related to this PR. Is it correct?

**Comment by mbluj** [comment]
FYI, I will be offline next two weeks and so I will not be able to follow this PR.

File: RecoTauTag/Configuration/python/tools/adaptToRunAtMiniAOD.py [hunk 1]
```diff
@@ -7,241 +7,293 @@
...
+				if name == 'qualityCuts':
+					builder.qualityCuts.primaryVertexSrc = 'offlineSlimmedPrimaryVertices'
+				elif name == 'pfCandSrc':
+					builder.pfCandSrc = 'packedPFCandidates'
+		# Adapt supported modifiers and remove unsupported ones
+		modifiersToRemove_ = cms.VPSet()
+		for mod in getattr(self.process,'combinatoricRecoTaus'+self.postfix).modifiers:
+			if mod.name.value() == 'elec_rej':
+				modifiersToRemove_.append(mod)
+				continue
```
Comment (slava77) [inline RecoTauTag/Configuration/python/tools/adaptToRunAtMiniAOD.py:160]: this part, to L168 is particularly heavy on `getattr(self.process,'combinatoricRecoTaus'+self.postfix)`, which apparently points to the same object in this context/range of lines (the worst is the last one where `getattr` is called in a loop.
Please replace with access to an explicit object.

File: RecoTauTag/Configuration/python/tools/adaptToRunAtMiniAOD.py [hunk 2]
```diff
@@ -7,241 +7,293 @@
...
+			getattr(self.process,'recoTauAK4Jets08RegionPAT'+self.postfix).pfCandAssocMapSrc = cms.InputTag('boostedTauSeedsPAT'+self.postfix, 'pfCandAssocMapForIsolation')
+			getattr(self.process,'recoTauPileUpVertices'+self.postfix).src = "offlineSlimmedPrimaryVertices"
+
+		for moduleName in self.miniAODTausTask.moduleNames():
+			self.convertModuleToMiniAODInput(moduleName)
+
+
+		# Adapt TauPiZeros producer
+		getattr(self.process,'ak4PFJetsLegacyHPSPiZeros'+self.postfix).builders[0].qualityCuts.primaryVertexSrc = "offlineSlimmedPrimaryVertices"
+		getattr(self.process,'ak4PFJetsLegacyHPSPiZeros'+self.postfix).jetSrc = jetCollection
```
Comment (slava77) [inline RecoTauTag/Configuration/python/tools/adaptToRunAtMiniAOD.py:131]: please reduce use of getattr to 1

File: RecoTauTag/Configuration/python/tools/helpers.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+## Helpers to perform some technically boring tasks like looking for all modules with a given parameter
+## and replacing that to a given value
+
+# Next two lines are for backward compatibility, the imported functions and
+# classes used to be defined in this file.
+from FWCore.ParameterSet.MassReplace import massSearchReplaceAnyInputTag, MassSearchReplaceAnyInputTagVisitor
+from FWCore.ParameterSet.MassReplace import massSearchReplaceParam, MassSearchParamVisitor, MassSearchReplaceParamVisitor
+
+class CloneTaskVisitor(object):
```
Comment (slava77) [inline RecoTauTag/Configuration/python/tools/helpers.py:11]: this looks like an extension to what's available in `PhysicsTools.PatAlgos.tools.helpers.CloneSequenceVisitor`
I think that this part is somewhat misplaced.
The simpler change is perhaps to add these generic helpers to `PhysicsTools.PatAlgos.tools.helpers`,
but eventually, some place in `FWCore.ParameterSet` would be more appropriate.

@cms-sw/core-l2 please check and advise if placement in `FWCore.ParameterSet` is reasonable
Thank you.

Comment (makortel) [reply RecoTauTag/Configuration/python/tools/helpers.py:11]: To me (e.g.) the `PhysicsTools.PatAlgos` (continues to) looks a reasonable place for these helpers.

**Comment by slava77** [comment]
> FYI, I will be offline next two weeks and so I will not be able to follow this PR.

@mbluj 
please clarify on the status of updating this PR.
Thank you.

File: RecoTauTag/Configuration/python/tools/helpers.py (hunk 3 above)
Comment (mbluj) [reply RecoTauTag/Configuration/python/tools/helpers.py:11]: Thanks for the advice, the helper will be moved.

**Comment by mbluj** [comment]
> > FYI, I will be offline next two weeks and so I will not be able to follow this PR.
> 
> @mbluj
> please clarify on the status of updating this PR.
> Thank you.

I am back to work and I plan to implement suggested changes today (10 Aug).

File: RecoTauTag/Configuration/python/tools/adaptToRunAtMiniAOD.py (hunk 2 above)
Comment (mbluj) [reply RecoTauTag/Configuration/python/tools/adaptToRunAtMiniAOD.py:131]: Done

File: RecoTauTag/Configuration/python/tools/adaptToRunAtMiniAOD.py (hunk 1 above)
Comment (mbluj) [reply RecoTauTag/Configuration/python/tools/adaptToRunAtMiniAOD.py:160]: Done

**Comment by Martin-Grunewald** [comment]
-1
please rebase to get rid of commits

**Comment by mbluj** [comment]
Rebased, conflicts should be fixed by it

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@cmsbuild please test

24 hrs seems too long for the last request

**Comment by slava77** [comment]
+reconstruction

for https://github.com/cms-sw/cmssw/pull/34576	717ee73 
- code changes are in line with the PR description and the follow up review
- jenkins tests pass and comparisons with the baseline show no differences
    - the updated parts of tau reco from mini are still experimental, they are not in the matrix workflows with comparisons
    - a unit test runtestRecoTauTagRecoTau includes `rerunTauRecoOnMiniAOD.py` and it passed (although it's not obvious if the boosted taus actually ran)

**Comment by perrotta** [comment]
+1

**Comment by perrotta** [comment]
merge

**Pull request merged**